lib/artifact.js

```diff
@@ -0,0 +1,109 @@
+    /**
+     * Instances of this class represent a Shelf artifact.
+     */
+    class Artifact {
+        /**
+         * @param {string} uri The full URL we're accessing.
+         * @param {string} authToken The token used to authenticate access.
+         */
+        constructor(uri, authToken) {
+            this.authToken = authToken;
+            this.metadata = new Metadata(uri, authToken);
+            this.uri = uri;
+        }
+
+
+        /**
+         * Uploads content to Shelf at the provided path.
+         *
+         * @param {(string|ReadStream)} content
+         * @return {Promise.<Object>}
+         */
+        upload(content) {
+            var options;
+
+            logger.debug(`In ${this.constructor.name}.upload()`);
+            options = requestOptions.createOptions(this.authToken);
+            options.url = this.uri;
+            options.formData = {
+                file: {
+                    value: content,
+                    options: {
+                        filename: "file",
+                        contentType: "application/octet-stream"
+                    }
+                }
+            };
+            logger.debug(`POST request options: ${JSON.stringify(options, null, 4)}`);
+
+            return requestPromise.post(options).then((response) => {
+                logger.debug(`In ${this.constructor.name}.upload() -> request.post().then()`);
+
+                return responseHandler.handleArtifactResponse(response).then((responseObj) => {
+                    return responseObj;
+                });
+            });
+        }
+
+
+        /**
+         * Uploads content to Shelf at the provided path plus a timestamp.
+         *
+         * @param {(string|ReadStream)} content
+         * @return {Promise.<Object>}
+         */
+        uploadWithTimestamp(content) {
+            var options;
+
+            logger.debug(`In ${this.constructor.name}.upload()`);
+            options = requestOptions.createOptions(this.authToken);
+            options.url = `${this.uri}/${dateService.now()}`;
+            options.formData = {
+                file: {
+                    value: content,
+                    options: {
+                        filename: "file",
+                        contentType: "application/octet-stream"
+                    }
+                }
+            };
+            logger.debug(`POST request options: ${JSON.stringify(options, null, 4)}`);
+
+            return requestPromise.post(options).then((response) => {
+                logger.debug(`In ${this.constructor.name}.upload() -> request.post().then()`);
+
+                return responseHandler.handleArtifactResponse(response).then((responseObj) => {
+                    return responseObj;
+                });
+            });
+        }
+
+
+        /**
+         * Downloads the artifact from the current path.
+         *
+         * @return {Promise.<Object>}
```

Comment:
Yeah, I've been thinking about this as well. I think we should have a chat about this. `request-promise` would load it all into memory. So yeah, we should probably allow the user to provide a location for where to download it.